pageserver/src/storage_sync/index.rs

```diff
@@ -257,14 +259,12 @@ impl RemoteTimeline {
 
     pub fn from_index_part(timeline_path: &Path, index_part: IndexPart) -> anyhow::Result<Self> {
         let metadata = TimelineMetadata::from_bytes(&index_part.metadata_bytes)?;
-        let default_metadata = &IndexLayerMetadata::default();
-
-        let find_metadata = |key: &RelativePath| -> LayerFileMetadata {
-            index_part
-                .layer_metadata
-                .get(key)
-                .unwrap_or(default_metadata)
-                .into()
+
+        let find_metadata = |key: &RelativePath| -> Option<LayerFileMetadata> {
+            let index_metadata = index_part.layer_metadata.get(key)?;
+            Some(LayerFileMetadata {
+                file_size: index_metadata.file_size?,
```

Comment:
The reason I dismissed this design is what will happen after adding:

```patch
 pub struct LayerFileMetadata {
     file_size: u64,
+    file_hash: FileHash,
 }
```

When upgrading from this size-only metadata to size+hash, would we be ready to let go of the existing size-only metadata in that case as well? It feels like it would be nice to have an additional state for being-loaded or being-downloaded `RemoteIndex` which is when this upgrade could take place, but this is only starting to dawn upon me now that I've wrestled the download changes [1](https://github.com/neondatabase/neon/pull/2582/commits/b9eee63dba0937693d0ed64d12f14479800e286b), [2](https://github.com/neondatabase/neon/pull/2582/commits/08c7cb38fd01ea7c23043ab7a817feb0a99b21cd) in.

Reply:
>  would we be ready to let go of the existing size-only metadata in that case as well?

Maybe this is something that I'm not understanding and most probably due to the fact we still have no place where we propagate old layer metadata to the new one.

In my imagination, we still have the layer file at some point, where we could calculate both.
As a middle ground, we can use something else instead of `Option`:
```rust
enum LayerMetadata {
    Complete(ProperLayerMetadata),
    Partial(IndexLayerMetadata),
    None,
}
```
but maybe that's an overcomplication already, need to see the upgrade place first...

Reply:
Just recapping our yesterday chat here, probably with some extras.

> Maybe this is something that I'm not understanding and most probably due to the fact we still have no place where we propagate old layer metadata to the new one.

During initial download is the place where we combine or merge the possibly old metadata into the current version: https://github.com/neondatabase/neon/pull/2582/commits/08c7cb38fd01ea7c23043ab7a817feb0a99b21cd -- To facilitate removing the `Option` from `LayerFileMetadata` during download, the type of `LayerFileMetadata` must be closer to `IndexLayerMetadata`.

My best guess for *how* this could happen is to complicate the type by which we are doing the download by, which I intend to investigate today. I was earlier thinking that `RemoteIndex` needs to change but perhaps the `Option<LayerFileMetadata>` as might work as well. However after attach we don't really need the Option<LayerFileMetadata>` either.

Reply:
Adding two variants for `RemoteIndex`:

```
pub enum RemoteTimeline {
    AwaitingDownload {
        timeline_layers: HashMap<PathBuf, IndexLayerMetadata>,
        missing_layers: HashMap<PathBuf, IndexLayerMetadata>,
        metadata: TimelineMetadata,
    },
    Ready {
        timeline_layers: HashMap<PathBuf, LayerFileMetadata>,
        missing_layers: HashMap<PathBuf, LayerFileMetadata>,
        metadata: TimelineMetadata,
    },
}
```

Then in `compare_local_and_remote_timeline`:
- `RemoteTimeline::AwaitingDownload` gets turned into `RemoteTimeline::Ready` iff all of the metadata matches up, and all files are present
    - Metadata "matches up" is a new addition compared to previous
    - Assumes that we can read all metadata during startup, which is an easy assumption with the layer file length
- After a successful download a new operation will have to do the similar "upgrade" from RemoteTimeline::AwaitingDownload

At later "runtime" after the tenant has become ready, this does pose some issues however, because we will still have to continually manage the "which RemoteTimeline variant is it". I can imagine two solutions for that:

1. Complicate `RemoteIndex::AwaitingDownload` by separating the "local files" and "remote not yet downloaded files" (which is also managed in `LayerDownload`
2. Move the RemoteTimeline enumerization split higher up, by having different types for "being attached timeline" and "attached timeline"

I have a feeling the "needing to juggle two types at runtime" can only be pulled upwards until we have complete separation for completely ready tenants and possibly incomplete tenants. Doing that seems counterproductive in light of already conflicting #2575 and the future direction of #2595. I haven't yet read either of these but I assume at least in #2595 runtime juggling of different states is requirement.

Reply:
Feels like we're on the right track: we now have a notion of remote and local timelines.

I'm thinking, that with on-demand downloads we start to abandon the holistic timeline approach: there will be `RemoteLayer` in https://github.com/neondatabase/neon/pull/2595, so timeline would not necessary be "either with all layers locally or will all layers remotely" as in the enum above, it will have a mix of both.

In our case, a layer could have full metadata (and that metadata will be in index.json) or could have incomplete/empty metadata that require downloading the layer, calculating and updating the info in the index and remotely.

I'm not sure that you have to invest heavily into this approach unless you see the way forward now: your version has some good roots in it and we will improve it later.

Reply:
To summarize: At least I cannot see a way out of needing to have `Option<u64>` at the `LayerFileMetadata`. If not there, that kind of "type state" needs to exist as long as we have a central place to store tenants in different states. Or perhaps it will be discovered the instant #2582 is merged.

At this point #2582 still needs some actual usage of `LayerFileMetadata`, namely in the `compare_local_and_remote_timeline` to decide to download truncated layer files for example. I'll continue towards that and we can circle back to this topic if good ideas come up.